### PR TITLE
Feature/onclick

### PR DESCRIPTION
**Pull request by akileng56**
- Validate context object
- Add onclick options
- Improve error messages

### CODE REVIEW

File: src/components/ImageViewerContainer.ts [hunk 1]
```diff
@@ -125,38 +132,72 @@ class ImageViewerContainer extends Component<ImageViewerContainerProps, ImageVie
 
     public static validateProps(props: ImageViewerContainerProps): string {
         let message = "";
+        if (props.source === "systemImage" && props.mxObject && !(props.mxObject.inheritsFrom("System.Image"))) {
```
Comment (Andries-Smit) [inline src/components/ImageViewerContainer.ts]: is this the correct check? inheritsFrom? What if you display an Image itself? will it give you an error. in that case have a look at https://apidocs.mendix.com/7/client/mendix_lib_MxMetaObject.html#isA

File: src/components/ImageViewer.ts [hunk 2]
```diff
@@ -46,7 +47,7 @@ class ImageViewer extends Component<ImageViewerProps, ImageViewerState> {
                 ref: this.props.getRef
             },
             DOM.img({
-                onClick: this.toggleLightBox,
+                onClick: this.props.onClickOption === "openFullScreen" ? this.toggleLightBox : this.props.onClick,
                 src: this.props.imageUrl,
```
Comment (Andries-Smit) [inline src/components/ImageViewer.ts]: `onError` we could something too. Please check with Danny if we should; do nothing, fail gracefully with some info, show an error show a default/fallback images.

File: src/components/ImageViewer.ts [hunk 3]
```diff
@@ -46,7 +47,7 @@ class ImageViewer extends Component<ImageViewerProps, ImageViewerState> {
                 ref: this.props.getRef
             },
             DOM.img({
-                onClick: this.toggleLightBox,
+                onClick: this.props.onClickOption === "openFullScreen" ? this.toggleLightBox : this.props.onClick,
```
Comment (Andries-Smit) [inline src/components/ImageViewer.ts]: this logic should be in your onClick function.

File: src/components/ImageViewerContainer.ts [hunk 4]
```diff
@@ -31,7 +33,7 @@ interface ImageViewerContainerState {
 
 type DataSource = "systemImage" | "urlAttribute" | "staticUrl" | "staticImage";
 type Units = "auto" | "pixels" | "percentage";
-type onClickOptions = "doNothing" | "openFullScreen";
+type onClickOptions = "doNothing" | "callMicroflow" | "showPage" | "openFullScreen";
 
 class ImageViewerContainer extends Component<ImageViewerContainerProps, ImageViewerContainerState> {
     private subscriptionHandle: number;
```
Comment (Andries-Smit) [inline src/components/ImageViewerContainer.ts]: `private imageViewer: HTMLElement;` rename to `imageViewerNode`

File: src/components/ImageViewerContainer.ts [hunk 5]
```diff
@@ -79,7 +83,10 @@ class ImageViewerContainer extends Component<ImageViewerContainerProps, ImageVie
```
Comment (Andries-Smit) [inline src/components/ImageViewerContainer.ts:85]: `add("widget-image-view-container");` change class name to  `widget-image-viewer` Also update the css

File: src/components/ImageViewerContainer.ts [hunk 6]
```diff
@@ -125,38 +132,72 @@ class ImageViewerContainer extends Component<ImageViewerContainerProps, ImageVie
```
Comment (Andries-Smit) [inline src/components/ImageViewerContainer.ts:136]: `resetSubscriptions` should also subscribe to its relevant attributes like `dynamicUrlAttribute`

File: src/components/ImageViewerContainer.ts [hunk 7]
```diff
@@ -125,38 +132,72 @@ class ImageViewerContainer extends Component<ImageViewerContainerProps, ImageVie
...
-        }
-        if (!mxObject && (this.props.source === "systemImage" || this.props.source === "urlAttribute")) {
-            this.setState({ imageUrl: "" });
+            return UrlHelper.getDynamicResourceUrl(mxObject.getGuid(), mxObject.get("changedDate") as number);
         }
         if (this.props.source === "staticUrl") {
-            this.setState({ imageUrl: this.props.urlStatic });
+            return this.props.urlStatic;
         }
         if (this.props.source === "staticImage") {
```
Comment (Andries-Smit) [inline src/components/ImageViewerContainer.ts]: above `if` statements could be `else if`

File: src/components/ImageViewer.ts [hunk 8]
```diff
@@ -62,11 +63,9 @@ class ImageViewer extends Component<ImageViewerProps, ImageViewerState> {
...
-            this.setState({
-                isOpen: !this.state.isOpen
-            });
-        }
+        this.setState({
+            isOpen: !this.state.isOpen
+        });
     }
 
     private getStyle(value: string | number, type: string): number | string {
```
Comment (Andries-Smit) [inline src/components/ImageViewer.ts]: if statements could be `else if` please add a comment that `When type is auto default browser styles applies`

**Review by Andries-Smit — CHANGES_REQUESTED** [review]

File: src/components/ImageViewerContainer.ts (hunk 4 above)
Comment (akileng56) [reply src/components/ImageViewerContainer.ts]: Changed

File: src/components/ImageViewerContainer.ts (hunk 5 above)
Comment (akileng56) [reply src/components/ImageViewerContainer.ts:85]: Changed

File: src/components/ImageViewerContainer.ts (hunk 6 above)
Comment (akileng56) [reply src/components/ImageViewerContainer.ts:136]: Done

File: src/components/ImageViewerContainer.ts (hunk 1 above)
Comment (akileng56) [reply src/components/ImageViewerContainer.ts]: Its much better to use IsA

File: src/components/ImageViewerContainer.ts (hunk 7 above)
Comment (akileng56) [reply src/components/ImageViewerContainer.ts]: Fixed

File: src/components/ImageViewer.ts (hunk 3 above)
Comment (akileng56) [reply src/components/ImageViewer.ts]: Done

File: src/components/ImageViewer.ts (hunk 2 above)
Comment (akileng56) [reply src/components/ImageViewer.ts]: I think having a default /fallback image might confuse the users.  they might think the image has loaded when it has not.

File: src/components/ImageViewer.ts (hunk 8 above)
Comment (akileng56) [reply src/components/ImageViewer.ts]: Fixed

**Pull request merged**